internal/controller/flow_cilium.go

```diff
@@ -0,0 +1,208 @@
+
+// discoverHubbleRelayAddress uses a kubernetes clientset in order to discover the address of hubble-relay within kube-system.
+func discoverHubbleRelayAddress(ctx context.Context, ciliumNamespace string, clientset kubernetes.Interface) (string, error) {
+	service, err := clientset.CoreV1().Services(ciliumNamespace).Get(ctx, hubble_relay_name, metav1.GetOptions{})
+	if err != nil {
+		return "", errors.New("hubble Relay service not found; disabling Cilium flow collection")
+	}
+
+	if len(service.Spec.Ports) == 0 {
+		return "", errors.New("hubble Relay service has no ports; disabling Cilium flow collection")
+	}
+
+	address := fmt.Sprintf("%s:%d", service.Spec.ClusterIP, service.Spec.Ports[0].Port)
+	return address, nil
+}
+
+// NewCollector connects to Hubble Relay, sets up an Observer client, and returns a new Collector using it.
+func newCollector(ctx context.Context, logger *zap.SugaredLogger, ciliumNamespace string) (*Collector, error) {
+	config, err := NewClientSet()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new client set: %w", err)
+	}
+	hubbleAddress, err := discoverHubbleRelayAddress(ctx, ciliumNamespace, config)
+	if err != nil {
+		return nil, err
+	}
+	// Adjust this address if needed
+	conn, err := grpc.NewClient(hubbleAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to Cilium Hubble Relay: %w", err)
+	}
+	hubbleClient := observer.NewObserverClient(conn)
+	return &Collector{logger: logger, client: hubbleClient}, nil
+}
+
+// convertLayer4 function converts a slice of flow.Layer4 objects to a slice of pb.Layer4 objects.
+func convertLayer4(l4 *flow.Layer4) *pb.Layer4 {
+	layer4 := &pb.Layer4{}
+	if l4 == nil {
+		return layer4
+	}
+	switch protocol := l4.Protocol.(type) {
+	case *flow.Layer4_TCP:
+		layer4.Protocol = &pb.Layer4_Tcp{
+			Tcp: &pb.TCP{
+				SourcePort:      protocol.TCP.SourcePort,
+				DestinationPort: protocol.TCP.DestinationPort,
+			},
+		}
+	case *flow.Layer4_UDP:
+		layer4.Protocol = &pb.Layer4_Udp{
+			Udp: &pb.UDP{
+				SourcePort:      protocol.UDP.SourcePort,
+				DestinationPort: protocol.UDP.DestinationPort,
+			},
+		}
+	case *flow.Layer4_ICMPv4:
+		layer4.Protocol = &pb.Layer4_Icmpv4{
+			Icmpv4: &pb.ICMPv4{
+				Type: protocol.ICMPv4.Type,
+				Code: protocol.ICMPv4.Code,
+			},
+		}
+	case *flow.Layer4_ICMPv6:
+		layer4.Protocol = &pb.Layer4_Icmpv6{
+			Icmpv6: &pb.ICMPv6{
+				Type: protocol.ICMPv6.Type,
+				Code: protocol.ICMPv6.Code,
+			},
+		}
+	case *flow.Layer4_SCTP:
+		layer4.Protocol = &pb.Layer4_Sctp{
+			Sctp: &pb.SCTP{
+				SourcePort:      protocol.SCTP.SourcePort,
+				DestinationPort: protocol.SCTP.DestinationPort,
+			},
+		}
+	default:
+	}
+
+	return layer4
+}
+
+// convertCiliumWorkflows function converts a slice of flow.Workload objects to a slice of pb.Workload objects.
+func convertCiliumWorkflows(workloads []*flow.Workload) []*pb.Workload {
+	protoWorkloads := []*pb.Workload{}
```

Comment:
Sure. What are the benefits of this?

Reply:
This saves memory and time.
Otherwise, every `append` into the slice results in the slice running out of capacity, which results in allocating a new slice and copying the elements over. So in the end, the code caused the allocation of `len(workloads)` slices.
By pre-allocating it with the right capacity from the start, you allocate only one slice.